tools/internal/flock/flock_test.go

```diff
@@ -0,0 +1,111 @@
+package flock
+
+import (
+	"fmt"
+	"github.com/rs/zerolog/log"
+	"github.com/stretchr/testify/assert"
+	"path"
+	"regexp"
+	"testing"
+	"time"
+)
+
+// Base interval for all sleeps/polling etc in this test.
+const baseInterval = 10 * time.Microsecond
+
+const lockRetryInterval = baseInterval
+const lockTimeout = 10_000 * lockRetryInterval
+
+// This is the best I could do
+func TestWithLockEnsuresConcurrentExecution(t *testing.T) {
+	numRoutines := 25
+
+	// should be small, we want every routine to get the lock in serial
+	// _without_ triggering a timeout
+	lockSleepTime := lockTimeout / 1000
+
+	type result struct {
+		err error
+		id  int
+	}
+
+	ch := make(chan result)
+	opts := testOptions(t)
+	lockOwner := -1
+
+	for i := 0; i < numRoutines; i++ {
+		id := i // Copy to local variable to prevent leaks
+		go func() {
+			err := WithLock(opts, func() error {
+				log.Info().Msgf("[%d] got lock", id)
+				if lockOwner != -1 {
+					return fmt.Errorf("[%d] another routine also owns the lock? %d", id, lockOwner)
+				}
+				lockOwner = id
+
+				log.Info().Msgf("[%d] sleeping for %s", id, lockSleepTime)
+				time.Sleep(lockSleepTime)
+
+				log.Info().Msgf("[%d] woke, releasing lock", id)
+				lockOwner = -1
+
+				return nil
+			})
+			ch <- result{err, id}
+		}()
+	}
+
+	// Verify none of the routines returned an error
+	for i := 0; i < numRoutines; i++ {
+		r := <-ch
```

Comment:
This is ultra nitpicky and feel free to ignore. In theory if you ever get a situation where something happens to one of your go routines and it isn't able to send a result on the channel this could deadlock and hang for ever. This is highly unlikely for this particular test, more intended as an FYI of possible gotchas.

Reply:
Thanks for the pointers in this and other comments! I've updated both of these tests to use wait groups and have added timeouts.